protocols/identify/src/lib.rs

```diff
@@ -139,32 +140,50 @@ impl<T: AddrManager> ServiceProtocol for IdentifyProtocol<T> {
         );
     }
 
-    fn connected(&mut self, service: &mut ServiceContext, session: &SessionContext, version: &str) {
+    fn connected(&mut self, service: &mut ServiceContext, session: &SessionContext, _ver: &str) {
         if session.remote_pubkey.is_none() {
             error!("IdentifyProtocol require secio enabled!");
             service.disconnect(session.id);
             self.secio_enabled = false;
             return;
         }
 
-        let remote_info = RemoteInfo::new(
-            session.clone(),
-            version,
-            Duration::from_secs(DEFAULT_TIMEOUT),
-        );
+        // Update listen address added after current protocol init.
+        let listens = service.listens();
+        if listens.len() > self.listens_length {
+            for addr in listens.iter().skip(self.listens_length) {
+                self.listen_addrs.insert(0, addr.clone());
+            }
+            self.listens_length = listens.len();
+        }
+
+        let remote_info = RemoteInfo::new(session.clone(), Duration::from_secs(DEFAULT_TIMEOUT));
         trace!("IdentifyProtocol sconnected from {:?}", remote_info.peer_id);
         self.remote_infos.insert(session.id, remote_info);
 
-        let listen_addrs: HashSet<Multiaddr> = self
-            .observed_addrs
-            .values()
-            .chain(self.listen_addrs.iter())
+        let listen_addrs: Vec<Multiaddr> = self
+            .listen_addrs
+            .iter()
+            .filter(|addr| {
+                multiaddr_to_socketaddr(addr)
+                    .map(|socket_addr| !socket_addr.ip().is_unspecified())
+                    .unwrap_or(true)
+            })
             .take(MAX_ADDRS)
             .cloned()
             .collect();
-        let data = IdentifyMessage::ListenAddrs(listen_addrs.into_iter().collect()).encode();
+        let data = IdentifyMessage::ListenAddrs(listen_addrs).encode();
         service.send_message(session.id, self.id, data);
-        let data = IdentifyMessage::ObservedAddr(session.address.clone()).encode();
+
+        let observed_addr = session
```

Comment:
Use ```addr.decapsulate("/p2p").unwrap_or_else(||addr);``` to remove p2p phase which is a significant way.
https://docs.rs/parity-multiaddr/0.2.0/parity_multiaddr/struct.Multiaddr.html#method.decapsulate

Reply:
`decapsulate()` use String as argument, which is dirty, I don't like it.

Reply:
Anyway `decapsulate` is a more efficient implementation, you can define a function to hide the invoke of `decapsulate`.

Reply:
I just want remove the p2p part, `decapsulate` is different, it may remove the `/ws` part